File: granola_share/sync.py
"""The server's own sync loop: ask Granola for new notes and queue them for the pipeline."""

from __future__ import annotations

import asyncio
import json
import time
import traceback
from dataclasses import dataclass, field
from datetime import date, datetime, timedelta, timezone

from .config import Config
from .granola import GranolaClient, Meeting
from .store import Store

FIRST_RUN_LOOKBACK_DAYS = 30  # free plan only exposes 30 days anyway
OVERLAP_DAYS = 2


@dataclass
class SyncReport:
    listed: int = 0
    new: int = 0
    queued: list[str] = field(default_factory=list)  # titles handed to the pipeline
    errors: list[str] = field(default_factory=list)


def _since(store: Store) -> date:
    last = store.get_state("last_sync")
    if last:
        return datetime.fromisoformat(last).date() - timedelta(days=OVERLAP_DAYS)
    return date.today() - timedelta(days=FIRST_RUN_LOOKBACK_DAYS)


async def sync_once(cfg: Config, client: GranolaClient, store: Store, log=print, on_queued=None) -> SyncReport:
    report = SyncReport()
    since = _since(store)
    async with client.session() as session:
        stubs = await client.list_meetings(session, since=since)
        report.listed = len(stubs)
        known = store.known_ids()
        new_ids = [m.id for m in stubs if m.id not in known]
        report.new = len(new_ids)
        log(f"[sync] {len(stubs)} meetings since {since}, {len(new_ids)} new")
        if not new_ids:
            store.set_state("last_sync", datetime.now(timezone.utc).isoformat())
            return report
        stub_by_id = {m.id: m for m in stubs}
        full: list[Meeting] = []
        try:
            full = await client.get_meetings(session, new_ids)
        except Exception as e:  # fall back to the stub content if the batch call fails
            report.errors.append(f"get_meetings: {e}")
            log(f"[sync] get_meetings failed, using list data: {e}")
        full_by_id = {m.id: m for m in full}
        for mid in new_ids:
            m = full_by_id.get(mid) or stub_by_id[mid]
            if not m.notes_markdown and mid in stub_by_id:
                m.notes_markdown = stub_by_id[mid].notes_markdown
            if cfg.include_transcripts and not m.transcript:
                m.transcript = await client.get_transcript(session, mid)
            try:
                store.enqueue(m)
                report.queued.append(m.title)
                log(f"[sync] queued '{m.title}'")
            except Exception as e:
                report.errors.append(f"{mid}: {e}")
                log(f"[sync] failed on {mid}: {e}\n{traceback.format_exc()}")
        _dump_debug(cfg, full or list(stub_by_id.values()))
    if report.queued and on_queued:
        on_queued()
    store.set_state("last_sync", datetime.now(timezone.utc).isoformat())
    return report


def _dump_debug(cfg: Config, meetings: list[Meeting]) -> None:
    """Keep the last raw payload around so we can adjust field mapping if Granola changes."""
    try:
        cfg.debug_dir.mkdir(parents=True, exist_ok=True)
        (cfg.debug_dir / "last_meetings.json").write_text(
            json.dumps([m.raw for m in meetings[:5]], indent=2)[:500_000]
        )
    except Exception:
        pass


def run_loop(cfg: Config, client: GranolaClient, store: Store, log=print, stop=None, on_queued=None) -> None:
    while True:
        try:
            asyncio.run(sync_once(cfg, client, store, log=log, on_queued=on_queued))
        except Exception as e:
            log(f"[sync] error: {e}")
        if stop is not None and stop.is_set():
            return
        for _ in range(cfg.poll_interval_seconds):
            if stop is not None and stop.is_set():
                return
            time.sleep(1)

File: engine/src/StudyStash.Core/Sync.cs
using System.Globalization;
using System.Text.Json.Nodes;

namespace StudyStash.Core;

public sealed class SyncReport
{
    public int Listed { get; set; }
    public int New { get; set; }
    public List<string> Queued { get; } = []; // titles handed to the pipeline
    public List<string> Errors { get; } = [];
}

/// <summary>The library's own sync (server_sync): ask Granola for new lectures and queue them for the pipeline.</summary>
public static class Sync
{
    public const int FirstRunLookbackDays = 30; // the free plan only shows 30 days anyway
    public const int OverlapDays = 2;

    /// <summary>Two days before the last sync, or 30 days back the first time.</summary>
    public static DateOnly Since(Store store)
    {
        string? last = store.GetState("last_sync");
        if (last is not null)
            return DateOnly.FromDateTime(DateTimeOffset.Parse(last, CultureInfo.InvariantCulture).DateTime).AddDays(-OverlapDays);
        return DateOnly.FromDateTime(DateTime.Today).AddDays(-FirstRunLookbackDays);
    }

    public static async Task<SyncReport> SyncOnceAsync(Config cfg, ILectureSource client, Store store, Action<string>? log = null,
        Action? onQueued = null, CancellationToken ct = default)
    {
        log ??= Console.WriteLine;
        var report = new SyncReport();
        var since = Since(store);
        await using (var session = await client.SessionAsync(ct))
        {
            var stubs = await client.ListMeetingsAsync(session, since, ct: ct);
            report.Listed = stubs.Count;
            var known = store.KnownIds();
            var newIds = stubs.Where(m => !known.Contains(m.Id)).Select(m => m.Id).ToList();
            report.New = newIds.Count;
            log($"[sync] {stubs.Count} meetings since {since:yyyy-MM-dd}, {newIds.Count} new");
            if (newIds.Count == 0)
            {
                store.SetState("last_sync", Py.IsoNowUtc());
                return report;
            }
            var stubById = new Dictionary<string, Meeting>();
            foreach (var m in stubs) stubById[m.Id] = m;
            var full = new List<Meeting>();
            try
            {
                full = await client.GetMeetingsAsync(session, newIds, ct);
            }
            catch (Exception e) when (!ct.IsCancellationRequested) // the list's own copy still has the basics
            {
                report.Errors.Add($"get_meetings: {e.Message}");
                log($"[sync] get_meetings failed, using list data: {e.Message}");
            }
            var fullById = new Dictionary<string, Meeting>();
            foreach (var m in full) fullById[m.Id] = m;
            foreach (string id in newIds)
            {
                var m = fullById.GetValueOrDefault(id) ?? stubById[id];
                if (m.NotesMarkdown.Length == 0 && stubById.TryGetValue(id, out var stub)) m.NotesMarkdown = stub.NotesMarkdown;
                if (cfg.IncludeTranscripts && m.Transcript.Length == 0) m.Transcript = await client.GetTranscriptAsync(session, id, ct);
                try
                {
                    store.Enqueue(m);
                    report.Queued.Add(m.Title);
                    log($"[sync] queued '{m.Title}'");
                }
                catch (Exception e)
                {
                    report.Errors.Add($"{id}: {e.Message}");
                    log($"[sync] failed on {id}: {e.Message}\n{e}");
                }
            }
            DumpDebug(cfg, full.Count > 0 ? full : stubById.Values.ToList());
        }
        if (report.Queued.Count > 0) onQueued?.Invoke();
        store.SetState("last_sync", Py.IsoNowUtc());
        return report;
    }

    /// <summary>Keep the last raw lectures, so the field mapping can be fixed if Granola changes its output.</summary>
    static void DumpDebug(Config cfg, List<Meeting> meetings)
    {
        try
        {
            Directory.CreateDirectory(cfg.DebugDir);
            var raws = new JsonArray(meetings.Take(5).Select(m => m.Raw.DeepClone()).ToArray());
            Py.WriteText(Path.Combine(cfg.DebugDir, "last_meetings.json"), Py.Head(PyJson.Dumps(raws, indent: 2), 500_000));
        }
        catch (Exception)
        {
            // only a debugging aid
        }
    }

    /// <summary>Sync, then wait poll_interval_seconds, until stopped. A failed round is logged and tried again.</summary>
    public static async Task RunLoopAsync(Config cfg, ILectureSource client, Store store, Action<string>? log = null,
        Action? onQueued = null, CancellationToken stop = default)
    {
        log ??= Console.WriteLine;
        while (!stop.IsCancellationRequested)
        {
            try
            {
                await SyncOnceAsync(cfg, client, store, log, onQueued, stop);
            }
            catch (Exception e) when (!stop.IsCancellationRequested)
            {
                log($"[sync] error: {e.Message}");
            }
            try
            {
                await Task.Delay(TimeSpan.FromSeconds(cfg.PollIntervalSeconds), stop);
            }
            catch (OperationCanceledException)
            {
                return;
            }
        }
    }
}
